add max14001 support

kseerp opened the pull request:
The MAX14001 is configurable, isolated 10-bit ADCs for multi-range binary inputs.

Datasheet: https://www.analog.com/media/en/technical-documentation/data-sheets/max14001-max14002.pdf

Signed-off-by: Kim Seer Paller <kimseer.paller@analog.com>

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml:
```diff
@@ -0,0 +1,61 @@
+    description: External reference, needed to establish input scaling.
+
+  spi-max-frequency:
+    maximum: 5000000
```
you should $ref the spi-peripheral yaml... Check [here](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/iio/addac/adi,ad74413r.yaml#L107) for an example

Note that you'll need to add `unevaluatedProperties: false` instead of `additionalProperties: false`

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml:
```diff
@@ -0,0 +1,61 @@
+        #size-cells = <0>;
+        status = "okay";
+
+        max140001: adc@0 {
```
just do `adc@0 {...`

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml:
```diff
@@ -0,0 +1,61 @@
+            compatible = "adi,max14001";
+            reg = <0>;
+            #address-cells = <1>;
+            #size-cells = <0>;
```
Drop `#address-cells` and ` #size-cells`. You don't have any child nodes

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+
+#define MAX14001_WRITE_REG(reg_addr, data) \
+	bitrev16((u16)(MAX14001_REG_ADDR_MASK(reg_addr) | \
+	((u16)data) | MAX14001_SET_WRITE_BIT))
```
can't you use FIELD_GET() or FIELD_PREP() in the above?

kseerp replied:
I had to configure the bit order of the Raspberry Pi controller I was using to transmit data in MSB-first format. Since the first bit clocked into the SDI port is D[0], the data LSB and usually, SPI masters clock out the data MSB first, so must reverse data prior to outputting it to the max14001 SDI pin. So with rpi, I needed to reverse the bit order before sending it to the SDI pin. After implementing this, I was able to test and verify the default register values.

Perhaps I'll try to modify and configure it in LSB format so no need to bit reverse.

nunojsa replied:
That sounds very odd... You control how data is set and you just pass in a buffer to the spi controller that just clocks bytes out of your buffer starting from 0. I would be very surprised if any spi controller would do something like this. Note that your driver should work in any platform not just RPI.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+			   &st->spi_tx_buffer);
+
+	return spi_write(st->spi, &st->spi_tx_buffer, 2);
+}
```
can't regmap() be used instead?

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	 * transfer buffers to live in their own cache lines.
+	 */
+	u16 spi_tx_buffer ____cacheline_aligned;
+	u16 spi_rx_buffer ____cacheline_aligned;
```
you just need the first buffer to be cacheline aligned... Also it looks like your types can actually be __be16

And `____cacheline_aligned` is not enough. You need to `__aligned(IIO_DMA_MINALIGN)`. Note that `IIO_DMA_MINALIGN` is still not available in our tree. Others have done it like [this](https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/addac/ad74115.c#L43)...

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+{
+	int ret;
+
+	put_unaligned_be16(MAX14001_READ_REG(reg_addr), &st->spi_tx_buffer);
```
This is not really an unaligned access AFAICT... It looks like you want cpu_to_be16() instead. But Looks like regmap can be used, so you can drop this functions and even your spi buffers

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+
+	ret = spi_read(st->spi, &st->spi_rx_buffer, 2);
+	if (ret < 0)
+		return ret;
```
Again, I think regmap() can be used so this comment does not matter much but just FYI, there is [spi_write_then_read()](https://elixir.bootlin.com/linux/latest/C/ident/spi_write_then_read)...

kseerp replied:
I am not yet certain if using the regmap would be appropriate. However, given the timing and bit arrangement required for transfer, I think it may be more applicable to use the SPI transfer method.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	int data;
+
+	/* Wait for 100ms */
+	msleep(100);
```
Note that `msleep(100);` is already self explanatory :). So this comment does not add any value. Now, why we are sleeping at all raises some questions to me... I would expect to see a sleep in here if some reset was being done on the device which I'm not seeing. Any reason for this sleep?

Speaking in reset, does this part has a reset pin? If so, that is something simple enough to support from the begining.

kseerp replied:
As suggested in the datasheet config flow chart (p. 30), after power up or after the module is loaded in this case, I added a 100ms delay after power up. Since the max14001pmod has no reset pin and relies on a software reset and power on reset, I plan to include the software power on-reset after the module is loaded and then follow the config flow chart accordingly

nunojsa replied:
Well, but in this case you have no idea about timings so you cannot just put a random sleep in here. And if we do a soft reset? In that case you are not resetting the chip anyways.

> Since the max14001pmod has no reset pin 

You are doing a driver for max14001 so it needs to be generic. You cannot "attach" yourself to pmod... That said, I do not see and RST pin on the datasheet.

>  I'm planning to include the software power on-reset after 

That is a must if you want to keep this sleep() in here...

pcercuei replied:
At the very least, the right place for the `msleep(100)` would be after the regulator is enabled.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+		return ret;
+
+	/* Enable SPI Registers Write */
+	max14001_write(st, MAX14001_WEN, MAX14001_WRITE_WEN);
```
there's no error handling for `max14001_write()`... Ditto for all the other calls.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	max14001_write(st, MAX14001_CFGV, 0x000);
+	max14001_write(st, MAX14001_ENBLV, 0x000);
+
+	/* Enable MV Flag */
```
What is "MV flag"?

kseerp replied:
this pertains to memory validation. as indicated on config flow chart, need to enable MV flag to allow MV error to assert a fault

nunojsa replied:
Then I would suggest to improve the comment. Use "Memory validation" instead of MV and explain what it is doing. Otherwise remove the comment since it does not really help much

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	max14001_write(st, MAX14001_WEN, MAX14001_WRITE_WEN);
+
+	/* Enable field side current sink */
+	max14001_write(st, MAX14001_ENBL, 0x010);
```
This looks like it could be a devicetree property rather than unconditionally enable it

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	/* Disable SPI Registers Write */
+	max14001_write(st, MAX14001_WEN, 0x000);
+
+	return 0;
```
this should be `return max14001_write()`

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	/* Disable SPI Registers Write */
+	max14001_write(st, MAX14001_WEN, 0x000);
+
+	/* Enable SPI Registers Write */
```
Alright, at this point you'll have to explain why there's this ping pong in "Disable SPI Registers Write" and "Enable SPI Registers Write"?

Maybe this deserves a comment...

kseerp replied:
i forgot to modify this part before creating the PR, i'm testing on something and i intended not to include this part for it reiterate the previous procedure

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	max14001_write(st, MAX14001_WEN, 0x000);
+
+	return 0;
+}
```
Nit: I would move this function as close as possible as the call site. So, right above the probe() function

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+		return IIO_VAL_FRACTIONAL;
+
+	case IIO_CHAN_INFO_OFFSET:
+		st->vref_uv = regulator_get_voltage(st->vref);
```
ditto...

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	st = iio_priv(indio_dev);
+	st->spi = spi;
+
+	spi_set_drvdata(spi, indio_dev);
```
drop this... Note `spi_get_drvdata()` is not being called anywhere. That's a good indicator you do not need this

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+	spi_set_drvdata(spi, indio_dev);
+	mutex_init(&st->lock);
+
+	indio_dev->dev.parent = &spi->dev;
```
drop this. Already done by the IIO core

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,301 @@
+static const struct of_device_id max14001_of_match[] = {
+	{ .compatible = "adi,max14001" },
+	{ }
+};
```
MODULE_DEVICE_TABLE(of, max14001_of_match)

nunojsa reviewed (commented):
There's still some work to be done on this one... 

Also note that I expect you to first send this driver upstream before merging it with our master tree.

kseerp commented:
Changelogs:

- changed types and added __aligned(IIO_DMA_MINALIGN) on buffers
- added software power on reset
- added error handling for every calls
- improved code comments
- moved functions close to the call site
- modified and used optional regulator
- included additional device configurations instead of relying on default settings

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	};
+
+	put_unaligned_be16(MAX14001_READ_REG(reg_addr),
+			   &st->spi_tx_buffer);
```
As stated, this is not an unaligned access... I'm also not sure what you are doing with the `bitrev16()` calls on the addr. AFAICS from the datasheet, the address should actually be an u8 (5bits addr + 1bit W/R).  Moreover, looking at the datasheet and how the SPI command goes on the line, it looks like the transfer is actually little endian ("The first bit clocked into the SDI port is D[0], the data LSB"? Hence, you need to change your spi words to __le16 instead of __be16

That said, this function together with the macros is a bit confusing. To me, this looks like it should be:

```
u16 tx = 0;

/* spi_transfer definition */

tx = FIELD_PREP(MAX14001_ADDR_MASK, ref_addr);
/* not that being your arch already little endian the below function won't do  nothing. But it will matter if your arch is big endian */
st->spi_tx_buffer = cpu_to_le16(tx);
....

// to fetch the data,
*val = __le16_to_cpu(st->spi_rx_buffer) & MAX14001_DATA_MASK;
// or, for correctness you could also have (but I don't think it really maters so I would probably go with option 1)
*val = FIELD_GET(MAX14001_DATA_MASK, __le16_to_cpu(st->spi_rx_buffer));
```

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+{
+	put_unaligned_be16(MAX14001_WRITE_REG(reg_addr, data),
+			   &st->spi_tx_buffer);
+
```
On the right side, this should be:

```
u16 tx;

tx = FIELD_PREP(MAX14001_ADDR_MASK, reg_addr);
/* for coherence, consider adding a define for BIT(10) */
tx |= FIELD_PREP(BIT(10), 1);
tx |= FIELD_PREP(MAX14001_DATA_MASK, data);

st->spi_tx_buffer = cpu_to_le16(tx);
....

```

Also note, there's no reason why `data is an integer`. Make it a `u16` please

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	} else {
+		dev_err(&st->spi->dev, "Failed to get adc property");
+		return -EINVAL;
+	}
```
Alright, this does not look fine... First, you should use the device agnostic API. So, this should simply be done like this:

```
st->use_fadc = device_property_read_bool(&st->spi->dev, "adi,use-adc");
```

ditto for all the other boolean properties

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	default:
+		dev_err(&st->spi->dev, "Invalid filter value");
+		return -EINVAL;
+	}
```
Ok, first thing is that `adi,filter` is not defined as a mandatory property in the bindings. That means, that users that don't set the property will fail to probe the device. Indeed, this does not feel like a mandatory property so that you should have a default value.

Secondly, you should make use of FIELD_PREP() instead of the switch case you have. So this could look like this:

```
if (!device_property_read_u32(&st->spi->dev, "adi,filter", &filter) {
    // note we only care if the property exists. You might still want to handle the case
    // where an actual error happened (rather than the property not existing) but typically no one ever cares :)
    // this 3 should have a define for it
    if (filter > 3)
        return dev_err_probe();
    
    // MAX14001_FILTER_MASK looks to be GENMASK(3, 2)
    reg_cfg |= FIELD_PREP(MAX14001_FILTER_MASK, filter);
}
```

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	if (of_property_read_bool(np, "current-source"))
+		reg_cfg |= BIT(4);
+	else
+		reg_cfg &= ~BIT(4);
```
If you initialize `reg_cfg = 0` at the declaration, then there's no bit to clear...

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	st = iio_priv(indio_dev);
+	st->spi = spi;
+
+	indio_dev->name = spi_get_device_id(spi)->name;
```
just use "max14001"

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	indio_dev->num_channels = ARRAY_SIZE(max14001_channels);
+	indio_dev->modes = INDIO_DIRECT_MODE;
+
+	st->vref = devm_regulator_get_optional(&spi->dev, "vref");
```
you can drop the regulator from your struct. You just need in here. What you need to save is `st->vref_mv`

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	if (ret)
+		return ret;
+
+	/* Power on reset */
```
this is more "Software reset..."

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	if (ret)
+		return ret;
+
+	msleep(100);
```
This should come after the reset, no?

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	ret = max14001_read(st, MAX14001_FLAGS, &data);
+	if (ret)
+		return ret;
+
```
So, you are still not explaining in the comments why you need this ping pong in "disabling SPI registers write" to read the flags? Note that "Read Flags" does not add nothing in here. "MAX14001_FLAGS" already tells me that :). Can't we just enable the spi register access in the beginning of the function and disable it in the end? 

But more important, you save the flags in `data` and then you don't do a things with it... If your goal is only to clear those flags, __that__ is something important to be in a comment

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+
+	/* Clear Faults */
+	for (i = MAX14001_FLTEN; i <= MAX14001_ACT; i++) {
+		ret = max14001_write(st, i, 0x000);
```
just write 0

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+
+	ret = max14001_cfg_reg(st);
+	if (ret)
+		return ret;
```
This is also far from ideal. So before calling `max14001_cfg_reg()` you disable the spi registers and then, after it, you start writing in the register. That means, you're hiding the enablement of the registers inside that function. That should, at least, deserve a comment. 

But, since you are defining another function to read the fw properties, it's a normal pattern to do it in the main probe() function before calling `max14001_config_init()`. You would need to cache the cfg_value and only write it in here. You should also rename the function to `max14001_fw_parse()`

nunojsa reviewed (commented):
Also make sure to fix your bindings. CI is failing to pass

kseerp commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,418 @@
+	 * DMA (thus cache coherency maintenance) requires the
+	 * transfer buffers to live in their own cache lines.
+	 */
+	__be16 spi_tx_buffer __aligned(IIO_DMA_MINALIGN) __nocast;
```
I'm not entirely sure why checkpatch is displaying a warning saying that "externs should be avoided in .c files, and the function definition argument IIO_DMA_MINALIGN should also have an identifier". However, placing __nocast seems to clear the warning.

nunojsa commented on Documentation/devicetree/bindings/iio/adc/adi,max14001.yaml:
```diff
@@ -0,0 +1,85 @@
+  - compatible
+  - reg
+  - spi-max-frequency
+  - adi,use-fadc
```
The above two look odd... Typically `spi-max-frequency` is not mandatory to have. But having `adi,use-fadc` definitely looks wrong to me. It is a boolean property and you are making it mandatory? Hence, how can you disable the feature? If you can't disable, it makes no sense to have it as a fw dts property.  If the default should be to have it enabled, you can flip the logic of the property. Make it something like, `adi,no-fadc`...

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+	 * DMA (thus cache coherency maintenance) requires the
+	 * transfer buffers to live in their own cache lines.
+	 */
+	__le16			spi_tx_buffer __aligned(IIO_DMA_MINALIGN) __nocast;
```
Forgot to comment on your comment. Please drop `__nocast`. It feels like it might be some issue with checkpatch... Note, no upstream driver is adding that attribute.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+		BIT(IIO_CHAN_INFO_SCALE) |
+		BIT(IIO_CHAN_INFO_OFFSET),
+	}
+};
```
nit: you could reduce the scope of this moving it closer to the place where it will be used.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+	struct device *dev = &st->spi->dev;
+	int filter;
+
+	st->cfg_value = 0;
```
No need to do this... Note that `devm_iio_device_alloc()` will allocate your struct and initialize all the elements to 0

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+
+	if (!device_property_read_u32(dev, "adi,filter", &filter)) {
+		if (filter > 3)
+			dev_err(dev, "Invalid filter value\n");
```
then you need to return `-EINVAL`

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+}
+
+static int max14001_read(struct max14001_state *st,
+			 unsigned int reg_addr, int *val)
```
val should also be u16 to be consistent with `max14001_write()`

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+	ret = max14001_read(st, MAX14001_FLAGS, &data);
+	if (ret)
+		return ret;
+
```
It's still not clear why you have all this calls to read the flags during the function? Is it needed really needed? Please add a comment saying why is it needed otherwise it is confusing for people reading the code.

nunojsa commented on drivers/iio/adc/max14001.c:
```diff
@@ -0,0 +1,365 @@
+
+	ret = max14001_write(st, MAX14001_CFGV, st->cfg_value | reg_cfg);
+	if (ret)
+		return ret;
```
It looks like you want to just update the bit in `st->cfg_value`. Doing the OR is not enough... Look at how [regmap_update_bits()](https://elixir.bootlin.com/linux/v6.3-rc6/source/drivers/base/regmap/regmap.c#L3248) does it. You can also have it as one liner.

Also, does the value read in MAX14001_CFG also applies to MAX14001_CFGV. Are these mirrored registers or something like that?

kseerp replied:
CFGV bits are continually compared to the CFG register. If any bits do not match, the memory validation bit in the FLAGS register is set and fault is asserted. When writing to configuration registers, verification registers have to be included as well.

nunojsa reviewed (changes requested):
Some more comments on my side. I think you can already take those comments and prepare a patchset to send upstream.

nunojsa commented:
@kseerp, what's the state of this PR? Do you have plans to upstream this?

kseerp commented:
> @kseerp, what's the state of this PR? Do you have plans to upstream this?

this has already been sent upstream. should this PR be closed for now?

nunojsa commented:
> this has already been sent upstream. should this PR be closed for now?

Was it accepted already? If so, please sync this PULL so we can merge the driver

kseerp commented:
It's not yet accepted. I will follow up on it, once accepted, I'll update this PR and let you know. Thanks.

kseerp commented:
> @kseerp not seeing this driver upstream... Any plan to continue that work or should I close this?

i agree to close this for now, i'll follow up on the sent patch series upstream when time allows

nunojsa commented:
> i agree to close this for now, i'll follow up on the sent patch series upstream when time allows

Based on this...
